hostmap.go

```diff
@@ -607,82 +475,26 @@ func (i *HostInfo) handshakeComplete(l *logrus.Logger) {
 		}
 	}
 
-	i.badRemotes = make([]*udpAddr, 0)
+	i.remotes.ResetBlockedRemotes()
 	i.packetStore = make([]*cachedPacket, 0)
 	i.ConnectionState.ready = true
 	i.ConnectionState.queueLock.Unlock()
 	i.ConnectionState.certState = nil
 }
 
-func (i *HostInfo) CopyRemotes() []*udpAddr {
-	i.RLock()
-	rc := make([]*udpAddr, len(i.Remotes), len(i.Remotes))
-	for x, addr := range i.Remotes {
-		rc[x] = addr.Copy()
-	}
-	i.RUnlock()
-	return rc
-}
-
 func (i *HostInfo) GetCert() *cert.NebulaCertificate {
 	if i.ConnectionState != nil {
 		return i.ConnectionState.peerCert
 	}
 	return nil
 }
 
-func (i *HostInfo) AddRemote(remote *udpAddr) *udpAddr {
-	if i.unlockedIsBadRemote(remote) {
-		return i.remote
-	}
-
-	for _, r := range i.Remotes {
-		if r.Equals(remote) {
-			return r
-		}
-	}
-
-	// Trim this down if necessary
-	if len(i.Remotes) > MaxRemotes {
-		i.Remotes = i.Remotes[len(i.Remotes)-MaxRemotes:]
-	}
-
-	rc := remote.Copy()
-	i.Remotes = append(i.Remotes, rc)
-	return rc
-}
-
 func (i *HostInfo) SetRemote(remote *udpAddr) {
-	i.remote = i.AddRemote(remote)
-}
-
-func (i *HostInfo) unlockedBlockRemote(remote *udpAddr) {
-	if !i.unlockedIsBadRemote(remote) {
-		// We copy here because we are taking something else's memory and we can't trust everything
-		i.badRemotes = append(i.badRemotes, remote.Copy())
+	// We copy here because we likely got this remote from a source that reuses the object
+	if !i.remote.Equals(remote) {
+		i.remote = remote.Copy()
+		i.remotes.LearnRemote(i.hostId, remote.Copy())
```

Comment:
do we need to call .Copy() twice here though? We could probably just reuse the `i.remote` copy in LearnRemote.

Reply:
If remote list doesn't have a discrete copy then any modification to `i.remote` will leak into the remote list. It's probably fine but I wanted to take the easy way out of the thought process here and keep them separated.